index.bs

```diff
@@ -1100,31 +1125,105 @@ strong map from handle ids to their corresponding objects.
 OwnershipModel = "root" / "none";
 </pre>
 
-<dfn>RemoteReference</dfn> represents a remote reference to an existing ECMAScript
-object in [=handle object map=] in the given [=Realm=].
+<dfn>RemoteReference</dfn> is either a <code>RemoteObjectReference</code>
+representing a remote reference to an existing ECMAScript object in
+[=handle object map=] in the given [=Realm=], or is a <code>SharedReference</code>
+representing a reference to a [=node=].
 
 <pre class="cddl remote-cddl local-cddl">
 Handle = text;
+SharedId = text;
+</pre>
 
-RemoteReference = {
+<pre class="cddl remote-cddl">
+RemoteReference = (
+  RemoteObjectReference //
+  SharedReference
+)
+
+RemoteObjectReference = {
    handle: Handle,
+   ?sharedId: SharedId
+   Extensible
+}
+
+SharedReference = {
+   ?handle: Handle,
+   sharedId: SharedId
    Extensible
 }
 </pre>
 
 Issue: handle "stale object reference" case.
 
+Note: if the provided reference has both <code>handle</code> and
+<code>sharedId</code>, the algorithm will ignore <code>handle</code> and respect
+only <code>sharedId</code>.
+
+### Deserialization ### {#data-types-reference-deserialization}
+
+<div algorithm>
+To <dfn>deserialize remote reference</dfn> given |remote reference|, |realm| and
+|session|:
+
+1. Assert |remote reference| matches the <code>RemoteReference</code> production.
+
+1. If |remote reference| matches the <code>SharedReference</code> production, return
+   [=deserialize shared reference=] given |remote reference|, |realm| and |session|.
+
+1. Assert |remote reference| matches the <code>RemoteObjectReference</code>
+   production.
+
+1. Return [=deserialize remote object reference=] given |remote reference| and
+   |realm|.
+
+</div>
+
 <div algorithm>
-To <dfn>deserialize remote reference</dfn> given |realm| and |remote reference|:
+To <dfn>deserialize remote object reference</dfn> given |remote object reference| and
+|realm|:
 
-1. Let |handle id| be the value of the <code>handle</code> field of |remote reference|.
+1. Let |handle id| be the value of the <code>handle</code> field of
+   |remote object reference|.
 
 1. Let |handle map| be |realm|'s [=handle object map=]
 
 1. If |handle map| does not contain |handle id|, then return [=error=] with
-   [=error code=] [=invalid argument=].
+   [=error code=] [=no such handle=].
 
-1. Return |handle map|[|handle id|].
+1. Return [=success=] with data |handle map|[|handle id|].
+
+</div>
+
+<div algorithm>
+To <dfn>deserialize shared reference</dfn> given |shared reference|, |realm| and
+|session|:
+
+1. Assert |shared reference| matches the <code>SharedReference</code> production.
+
+1. Let |browsing context| be [=get the browsing context=] with |realm|.
+
+1. If |browsing context| is <code>null</code>, return [=error=] with [=error code=]
```

Comment:
I changed this to return `no such node` since it's at least true that the node you're trying to access doesn't existing in the realm you're trying to access it in.

Reply:
We could make up a new error, or we could say "well you tried to look up a node in this realm and it wasn't found (because this realm doesn't have any nodes in it)". I don't really have a good sense for when people prefer to construct new errors vs when they want to reuse existing ones.

Reply:
Note that in WebDriver classic we raise a different error which is `no such window` if the related browsing context cannot be found. This is outside of the deserialization and serialization code but at least gives more details why the element / ShadowRoot wasn't found. 

With BiDi we basically operate on any browsing context and would already throw a `no such frame` error if eg. a script should be evaluated for a context / realm that doesn't exist. So maybe we are fine with keeping a single additional error then.

Reply:
> So this actually happens if |realm| doesn't have a `Window` object as its global scope (e.g. it's a worker realm). So I think we need a better error type.

I guess if the |realm| doesn't have a browsing context for whatever reason, we can consider it doesn't have a `frame`, and `[=no such frame=]` seems to be a proper error code.